nixos/modules/services/web-servers/ttyd.nix

```diff
@@ -0,0 +1,202 @@
+        };
+        description = ''
+          Attribute set of client options for xtermjs.
+          https://xtermjs.org/docs/api/terminal/interfaces/iterminaloptions/
+        '';
+      };
+
+      terminalType = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        example = "xterm-256color";
+        description = "Terminal type to report, default: xterm-256color.";
+      };
+
+      checkOrigin = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Do not allow websocket connection from different origin.";
+      };
+
+      maxClients = mkOption {
+        type = types.nullOr types.int;
+        default = null;
+        example = 0;
+        description = "Maximum clients to support (default: 0, no limit)";
+      };
+
+      indexFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = "Custom index.html path";
+      };
+
+      enableIPv6 = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Whether or not to enable IPv6 support.";
+      };
+
+      enableSSL = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Whether or not to enable SSL (https) support.";
+      };
+
+      certFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = "SSL certificate file path.";
+      };
+
+      keyFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = "SSL key file path.";
+      };
+
+      caFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = "SSL CA file path for client certificate verification.";
+      };
+
+      logLevel = mkOption {
+        type = types.nullOr types.int;
+        default = null;
+        example = 7;
+        description = "Set log level (default: 7).";
+      };
+    };
+  };
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    warnings = optional (cfg.password != null) "ttyd passwords will be stored as plaintext in the Nix store!";
+
+    assertions =
+      [ { assertion = cfg.enableSSL == true
+            -> cfg.certFile != null && cfg.keyFile != null && cfg.caFile != null;
+          message = "SSL is enabled for ttyd, but no certFile, keyFile or caFile has been specefied."; }
+        { assertion = ! (cfg.interface != null && cfg.socket != null);
+          message = "Cannot set both interface and socket for ttyd."; }
+        { assertion = (cfg.username != null -> cfg.password != null) &&
+                      (cfg.password != null -> cfg.username != null);
+          message = "Need to set both username and password"; }
+      ];
+
+    systemd.services.ttyd = {
+      description = "ttyd Web Server Daemon";
+
+      wantedBy = [ "multi-user.target" ];
+      requires = [ "sshd.service" ];
+      after = [ "sshd.service" ];
```

Comment:
By default, `sshd` is not enabled, so you should consider this here (does it need to be enabled? Then assertion)

Reply:
Good catch, not really. Thanks!.
I based this of of `shellinabox` so it just stayed there.

[1] https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-servers/shellinabox.nix